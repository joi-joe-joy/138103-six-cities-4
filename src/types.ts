export enum HouseType {
  APARTMENT = `apartment`,
  ROOM = `room`,
  HOUSE = `house`,
  HOTEL = `hotel`
};

export enum PlaceCardType {
  CITIES = `cities`,
  NEAR = `near`,
  FAVORITES = `favorites`,
  PROPERTY = `property`
};

export enum PageType {
  PROPERTY = `property`,
  MAIN = `main`,
  LOGIN = `login`,
  FAVORITES = `favorites`,
  FAVORITES_EMPTY = `favorites_empty`
};

export enum SortType {
  POPULAR = `popular`,
  TO_HIGH = `to-high`,
  TO_LOW = `to-low`,
  TOP_RATED = `top-rated`
};

export enum AuthStatus {
  NO_AUTH = `NO_AUTH`,
  AUTH = `AUTH`,
};

export interface Location {
  latitude: number,
  longitude: number,
  zoom: number
}

export interface City {
  location: Location,
  name: string
}

export interface User {
  id: number,
  avatarUrl: string,
  name: string,
  isPro: boolean
}

export type AuthInfo = User & {
  email: string
}

export interface Offer {
  id: number,
  title: string,
  maxAdults: number,
  description: string,
  isPremium: boolean,
  isFavorite: boolean,
  images: Array<string>,
  previewImage: string,
  goods: Array<string>,
  price: number,
  rating: number,
  bedrooms: number,
  type: HouseType.APARTMENT | HouseType.ROOM | HouseType.HOUSE | HouseType.HOTEL,
  host: User,
  location: Location,
  city: City,
};

export interface Comment {
  comment: string,
  date: string,
  id: number,
  rating: number,
  user: User
}
