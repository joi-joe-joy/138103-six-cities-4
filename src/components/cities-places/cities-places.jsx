import React from 'react';
import {connect} from "react-redux";
import {PlaceCardType} from "../../const";
import SortVariants from "../sort-variants/sort-variants";
import PlacesList from "../places-list/places-list";
import withActiveItem from "../../hocs/with-active-item/with-active-item";
import withSort from "../../hocs/with-sort/with-sort";
import {sortOffers} from "../../selector";
import pt from "prop-types";

const SortVariantsWrap = withSort(SortVariants);
const PlacesListWrap = withActiveItem(PlacesList);

const CitiesPlaces = (props) => {
  const {offers, currentCity} = props;

  return (
    <section className="cities__places places">
      <h2 className="visually-hidden">Places</h2>
      <b className="places__found">{offers.length} places to stay in {currentCity}</b>
      <SortVariantsWrap/>
      <PlacesListWrap
        type={PlaceCardType.CITIES}
        offers={offers}
      />
    </section>
  );
};

CitiesPlaces.propTypes = {
  offers: pt.array.isRequired,
  currentCity: pt.string.isRequired
};

const mapStateToProps = (state) => ({
  currentCity: state.city,
  offers: sortOffers(state.sorting, state.offersCityList)
});

export {CitiesPlaces};
export default connect(mapStateToProps)(CitiesPlaces);
