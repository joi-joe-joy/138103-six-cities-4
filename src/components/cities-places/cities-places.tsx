import * as React from 'react';
import {connect} from "react-redux";
import {PlaceCardType, Offer, City} from "../../types";
import SortVariants from "../sort-variants/sort-variants";
import PlacesList from "../places-list/places-list";
import withActiveItem from "../../hocs/with-active-item/with-active-item";
import withSort from "../../hocs/with-sort/with-sort";
import {getCity} from "../../reducer/data/selectors";
import {getSortedOffers} from "../../reducer/data/selectors";

interface Props {
  offers: Offer[];
  currentCity: City;
}

const SortVariantsWrap = withSort(SortVariants);
const PlacesListWrap = withActiveItem(PlacesList);

const CitiesPlaces: React.FC<Props> = (props: Props) => {
  const {offers, currentCity} = props;

  return (
    <section className="cities__places places">
      <h2 className="visually-hidden">Places</h2>
      <b className="places__found">{offers.length} places to stay in {currentCity.name}</b>
      <SortVariantsWrap/>
      <PlacesListWrap
        type={PlaceCardType.CITIES}
        offers={offers}
      />
    </section>
  );
};

const mapStateToProps = (state) => ({
  currentCity: getCity(state),
  offers: getSortedOffers(state)
});

export {CitiesPlaces};
export default connect(mapStateToProps)(CitiesPlaces);
