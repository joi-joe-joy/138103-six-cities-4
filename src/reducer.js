import {extend} from "./utils.js";
import offers from "./mocks/offers.js";

let initCity = offers[0].city;
let initOffers = offers.filter((offer) => offer.city === initCity);

// Получаем все значения городов
let cities = offers.map((offer)=>offer.city);
// Получаем уникальные значения городов
let citiesList = Array.from(new Set(cities));

const initialState = {
  city: initCity,
  offersCityList: initOffers,
  offers,
  citiesList
};

const ActionType = {
  CHANGE_CITY: `CHANGE_CITY`,
  GET_OFFERS_CITY_LIST: `GET_OFFERS_CITY_LIST`,
};

const ActionCreator = {
  changeCity: (city) => {
    let changedCity = city ? city : offers[0].city;

    return {
      type: ActionType.CHANGE_CITY,
      payload: changedCity
    };
  },
  getOffersCityList: (city) => {
    let offersCityList = offers.filter((offer) => offer.city === city);

    return {
      type: ActionType.GET_OFFERS_CITY_LIST,
      payload: offersCityList
    };
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_CITY:
      return extend(state, {
        city: action.payload
      });
    case ActionType.GET_OFFERS_CITY_LIST:
      return extend(state, {
        offersCityList: action.payload
      });
  }

  return state;
};

export {reducer, ActionType, ActionCreator};
