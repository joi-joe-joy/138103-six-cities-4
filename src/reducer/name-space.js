export const NameSpace = {
  DATA: `DATA`,
  USER: `USER`,
  PLACE: `PLACE`
};
