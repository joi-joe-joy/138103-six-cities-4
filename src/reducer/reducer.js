import {combineReducers} from "redux";
import {reducer as data} from "./data/data";
import {reducer as place} from "./place/place";
import {reducer as user} from "./user/user";
import {reducer as comments} from "./comments/comments";
import {reducer as favorite} from "./favorite/favorite";
import {NameSpace} from "./name-space";

export default combineReducers({
  [NameSpace.DATA]: data,
  [NameSpace.PLACE]: place,
  [NameSpace.USER]: user,
  [NameSpace.COMMENTS]: comments,
  [NameSpace.FAVORITE]: favorite,
});
