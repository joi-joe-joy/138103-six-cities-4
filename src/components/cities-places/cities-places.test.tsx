import * as React from "react";
import * as renderer from "react-test-renderer";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import {CitiesPlaces} from "./cities-places";
import {NameSpace} from "../../reducer/name-space";
import {BrowserRouter} from "react-router-dom";
import {Offer, HouseType} from "../../types";

const mockStore = configureStore([]);

const offers: Offer[] = [{
  bedrooms: 2,
  city: {
    name: `Paris`,
    location: {
      latitude: 48.85661,
      longitude: 2.351499,
      zoom: 13
    }
  },
  description: `Discover daily local life in city center, friendly neighborhood, clandestine casino, karaoke, old-style artisans, art gallery and artist studio downstairs.`,
  goods: [`Air conditioning`, `Laptop friendly workspace`, `Baby seat`, `Fridge`, `Breakfast`, `Washer`, `Washing machine`, `Dishwasher`, `Coffee machine`, `Towels`],
  host: {id: 25, name: `Angelina`, isPro: true, avatarUrl: `img/avatar-angelina.jpg`},
  id: 1,
  images: [`https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg`],
  isFavorite: false,
  isPremium: true,
  location: {
    latitude: 48.865610000000004,
    longitude: 2.350499,
    zoom: 16
  },
  maxAdults: 8,
  previewImage: `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/5.jpg`,
  price: 397,
  rating: 3.6,
  title: `Penthouse, 4-5 rooms + 5 balconies`,
  type: HouseType.HOTEL
}];

it(`Render CitiesPlaces correctly`, () => {
  const store = mockStore({
    [NameSpace.DATA]: {
      city: {
        name: `Paris`,
        location: {
          latitude: 48.85661,
          longitude: 2.351499,
          zoom: 13
        }
      },
      offers
    },
    [NameSpace.USER]: {
      authorizationStatus: `AUTH`
    }
  });
  const currentCity = offers[0].city;

  const tree = renderer
    .create(
        <Provider store={store}>
          <BrowserRouter>
            <CitiesPlaces
              currentCity={currentCity}
              offers={offers}
            />
          </BrowserRouter>
        </Provider>
    ).toJSON();

  expect(tree).toMatchSnapshot();
});
