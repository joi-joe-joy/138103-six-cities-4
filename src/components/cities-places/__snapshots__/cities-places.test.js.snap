// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render CitiesPlaces correctly 1`] = `
<section
  className="cities__places places"
>
  <h2
    className="visually-hidden"
  >
    Places
  </h2>
  <b
    className="places__found"
  >
    2
     places to stay in 
    Paris
  </b>
  <form
    action="#"
    className="places__sorting"
    method="get"
  >
    <span
      className="places__sorting-caption"
    >
      Sort by
    </span>
    <span
      className="places__sorting-type"
      onClick={[Function]}
      tabIndex="0"
    >
      Popular
      <svg
        height="4"
        style={
          Object {
            "marginBottom": "1px",
            "marginLeft": "5px",
          }
        }
        width="7"
      />
    </span>
    <ul
      className="places__options places__options--custom"
    >
      <li
        className="places__option places__option--active"
        onClick={[Function]}
        tabIndex="0"
      >
        Popular
      </li>
      <li
        className="places__option"
        onClick={[Function]}
        tabIndex="0"
      >
        Price: low to high
      </li>
      <li
        className="places__option"
        onClick={[Function]}
        tabIndex="0"
      >
        Price: high to low
      </li>
      <li
        className="places__option"
        onClick={[Function]}
        tabIndex="0"
      >
        Top rated first
      </li>
    </ul>
  </form>
  <div
    className="places__list cities__places-list tabs__content"
  >
    <article
      className="place-card cities__place-card"
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <div
        className="place-card__mark"
      >
        <span>
          Premium
        </span>
      </div>
      <div
        className="place-card__image-wrapper cities__image-wrapper"
      >
        <a
          href="#"
        >
          <img
            alt="Canal View Prinsengracht"
            className="place-card__image"
            height="200"
            src="img/apartment-01.jpg"
            width="260"
          />
        </a>
      </div>
      <div
        className="place-card__info"
      >
        <div
          className="place-card__price-wrapper"
        >
          <div
            className="place-card__price"
          >
            <b
              className="place-card__price-value"
            >
              €
              280
            </b>
            <span
              className="place-card__price-text"
            >
              / night
            </span>
          </div>
          <button
            className="place-card__bookmark-button button"
            onClick={[Function]}
            type="button"
          >
            <svg
              height="18"
              width="17"
            />
            <span
              className="visually-hidden"
            >
              To bookmarks
            </span>
          </button>
        </div>
        <div
          className="place-card__rating rating"
        >
          <div
            className="place-card__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "80%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <h2
          className="place-card__name"
        >
          <a
            onClick={[Function]}
          >
            Canal View Prinsengracht
          </a>
        </h2>
        <p
          className="place-card__type"
        >
          House
        </p>
      </div>
    </article>
    <article
      className="place-card cities__place-card"
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <div
        className="place-card__image-wrapper cities__image-wrapper"
      >
        <a
          href="#"
        >
          <img
            alt="Nice, cozy, warm big bed apartment"
            className="place-card__image"
            height="200"
            src="img/apartment-01.jpg"
            width="260"
          />
        </a>
      </div>
      <div
        className="place-card__info"
      >
        <div
          className="place-card__price-wrapper"
        >
          <div
            className="place-card__price"
          >
            <b
              className="place-card__price-value"
            >
              €
              80
            </b>
            <span
              className="place-card__price-text"
            >
              / night
            </span>
          </div>
          <button
            className="place-card__bookmark-button button"
            onClick={[Function]}
            type="button"
          >
            <svg
              height="18"
              width="17"
            />
            <span
              className="visually-hidden"
            >
              To bookmarks
            </span>
          </button>
        </div>
        <div
          className="place-card__rating rating"
        >
          <div
            className="place-card__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "80%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <h2
          className="place-card__name"
        >
          <a
            onClick={[Function]}
          >
            Nice, cozy, warm big bed apartment
          </a>
        </h2>
        <p
          className="place-card__type"
        >
          Private Room
        </p>
      </div>
    </article>
  </div>
</section>
`;
